Fix Lyric temperature setting when off

nprez83 opened the pull request:
In PR #67018 I added an if statement to prevent the target_temperature to be transmitted if the HVAC was off, but I didn't do that for target_temperature_low and target_temperature_high, which means if you shut the system off from auto mode, change the temperature setting in the front end, and then turn the system on, the temperature setting will have been changed. I added two additional if statement to prevent that, now if the temperature is changed in any way when the system if off, it will have no effect.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
None


## Proposed change
I added two if statement to prevent target_temperature_low and target_temperature_high from being transmitted when the device is off. 

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
- This PR is related to PR #67018 

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

nprez83 committed:
Fix Lyric Climate when Off

In PR #67018 I added an if statement to prevent the target_temperature to be transmitted if the HVAC was off, but I didn't do that for target_temperature_low and target_temperature_high, which means if you shut the system off from auto mode, change the temperature setting in the front end, and then turn the system on, the temperature setting will have been changed. I added two additional if statement to prevent that, now if the temperature is changed in any way when the system if off, it will have no effect.

thecode commented on homeassistant/components/lyric/climate.py:
```diff
@@ -223,16 +223,26 @@ def target_temperature(self) -> float | None:
         device = self.device
-        if device.changeableValues.autoChangeoverActive:
-            return device.changeableValues.coolSetpoint
+        if (
+            device.changeableValues.autoChangeoverActive
+            and HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF
+        ):
+            if device.changeableValues.autoChangeoverActive:
+                return device.changeableValues.coolSetpoint
+            return None
```
It would be cleaner to write this code like this:
```suggestion
        if (
            device.changeableValues.autoChangeoverActive
            and HVAC_MODES[device.changeableValues.mode] == HVAC_MODE_OFF
        ):
            return None
        if device.changeableValues.autoChangeoverActive:
            return device.changeableValues.coolSetpoint
            
```

nprez83 replied:
I agree, thanks for the suggestion, will make the changes now.

thecode replied:
In general it is better to reduce indented code if possible by using guard clause: https://deviq.com/design-patterns/guard-clause

I agree that there is a change in behavior in my suggestion, but I had in mind that you want to return `None` only in the case that `autoChangeoverActive` is active and HVAC mode is OFF.

if that is not the case you can also change the if to something like:
```python
        if (
            not device.changeableValues.autoChangeoverActive
            or HVAC_MODES[device.changeableValues.mode] == HVAC_MODE_OFF
        ):
            return None        
``` 
and in this case you don't need the additonal `if device.changeableValues.autoChangeoverActive` and you can return `device.changeableValues.coolSetpoint` after the `if`.

If my suggestion above does not match the behavior of the device, try to rewrite the code part by doing a first if and return. There are also two `if` one after another (which I missed on the first review) so at least these could be consolidated under one main `if` e.g:
```python
        if (
            device.changeableValues.autoChangeoverActive
            and HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF
        ):
            return device.changeableValues.coolSetpoint
```

nprez83 replied:
Thanks so much for the pointers. I have a background in coding but I'm new to python, so this really helps. I really like the first option, so I'll go with that. Thanks again!

thecode commented on homeassistant/components/lyric/climate.py:
```diff
@@ -223,16 +223,26 @@ def target_temperature(self) -> float | None:
         device = self.device
-        if device.changeableValues.autoChangeoverActive:
-            return device.changeableValues.heatSetpoint
+        if (
```
Same as above

nprez83 replied:
Will do

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -223,16 +223,26 @@ def target_temperature(self) -> float | None:
+        if (
+            device.changeableValues.autoChangeoverActive
+            and HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF
+        ):
+            if device.changeableValues.autoChangeoverActive:
+                return device.changeableValues.coolSetpoint
+            return None
```
You can merge this logic since the fall-through behavior is already returning None:
```suggestion
            and device.changeableValues.autoChangeoverActive
        ):
            return device.changeableValues.coolSetpoint
```

nprez83 replied:
Thank you. I think the changes I just committed make the logic much more concise, but happy to hear your thoughts.

nprez83 replied:
Thanks for suggestion @balloob, I think I missed your point initially, but I just made some additional changes to the target_temperature, target_temperature_high, and target_temperature_low logic that I think make them slightly more concise.

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -223,16 +223,26 @@ def target_temperature(self) -> float | None:
+        if (
+            device.changeableValues.autoChangeoverActive
+            and HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF
+        ):
+            if device.changeableValues.autoChangeoverActive:
+                return device.changeableValues.heatSetpoint
+            return None
```
Same.
```suggestion
            and device.changeableValues.autoChangeoverActive
        ):
            return device.changeableValues.heatSetpoint
```

nprez83 replied:
see above

nprez83 committed:
Update climate.py

I realized the if statements in the target_temperature, target_temperature_low, and target_temperature_high properties only prevent the set temperatures from being displayed in the frontend, but don't actually prevent the target temperature from being transmitted to the async_set_temperature function. To accomplish that, I added an if statement at the start of async_set_temperature, which returns None if the mode if OFF. This is what @MartinHjelmare had suggested in PR #67018 but I unfortunately misunderstood him.

nprez83 commented:
I realized the if statements in the target_temperature, target_temperature_low, and target_temperature_high properties only prevent the set temperatures from being displayed in the frontend, but don't actually prevent the target temperature from being transmitted to the async_set_temperature function. To accomplish that, I added an if statement at the start of async_set_temperature, which returns None if the mode if OFF. This is what @MartinHjelmare had suggested in PR #67018 but I unfortunately misunderstood him.

nprez83 committed:
Update climate.py

Made the logic for the target_temperature, target_temperature_high, and target_temperature_low slightly more concise.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -217,23 +217,26 @@ def target_temperature(self) -> float | None:
+            and HVAC_MODES[device.changeableValues.mode] != HVAC_MODE_OFF
+        ):
             return device.changeableValues.coolSetpoint
-        return None
```
We need to return a value explicitly everywhere if we return a value anywhere in a function. Please keep this.

nprez83 replied:
Roger that. Will change back. Thanks for the guidance.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -269,6 +272,9 @@ def max_temp(self) -> float:
     async def async_set_temperature(self, **kwargs) -> None:
         """Set new target temperature."""
+        if self.hvac_mode == HVAC_MODE_OFF:
+            return None
```
```suggestion
            return
```

nprez83 replied:
Thanks, will incorporate.

nprez83 committed:
Update climate.py

Explicitly returning none for the target_temperature properties, and changed the return statement when HVAC is OFF within async_set_temperature to return instead of return None.
